examples/structured_data/structured_data_classification_with_wide_deep_and_cross_networks.py

```diff
@@ -0,0 +1,424 @@
+                encoded_feature = onehot_encoder(value_index)
+        else:
+            # Use the numerical features as-is.
+            encoded_feature = tf.expand_dims(inputs[feature_name], -1)
+
+        encoded_features.append(encoded_feature)
+
+    all_features = layers.concatenate(encoded_features)
+    return all_features
+
+
+"""
+## Experiment 1: a baseline model
+
+In the first experiment, let's create a multi-layer feed-forward network, where the categorical features are
+one-hot encoded.
+"""
+
+
+def create_baseline_model():
+    inputs = create_model_inputs()
+    features = encode_inputs(inputs)
+
+    for units in hidden_units:
+        features = layers.Dense(units)(features)
+        features = layers.BatchNormalization()(features)
+        features = layers.ReLU()(features)
+        features = layers.Dropout(dropout_rate)(features)
+
+    outputs = layers.Dense(units=NUM_CLASSES, activation="softmax")(features)
+    model = keras.Model(inputs=inputs, outputs=outputs)
+    return model
+
+
+baseline_model = create_baseline_model()
+keras.utils.plot_model(baseline_model, show_shapes=True)
+
+run_experiment(baseline_model)
+
+"""
+The baseline linear model achieves ~76.4% test accuracy.
+"""
+
+"""
+## Experiment 2: wide and deep model
+
+In the second experiment, we create a wide and deep model. The wide part of the model a linear model,
+while the deep part of the model is a multi-layer feed-forward network.
+
+Use the sparse representation of the input features in the wide part of the model and the
+dense representation of the input features for the deep part of the model.
+
+Note that every input features contributes to both parts of the model with different representations.
+"""
+
+
+def create_wide_and_deep_model():
+
+    inputs = create_model_inputs()
+    wide = encode_inputs(inputs)
+    wide = layers.BatchNormalization()(wide)
+
+    deep = encode_inputs(inputs, use_embedding=True)
+    for units in hidden_units:
+        deep = layers.Dense(units)(deep)
+        deep = layers.BatchNormalization()(deep)
+        deep = layers.ReLU()(deep)
+        deep = layers.Dropout(dropout_rate)(deep)
+
+    merged = layers.concatenate([wide, deep])
+    outputs = layers.Dense(units=NUM_CLASSES, activation="softmax")(merged)
+    model = keras.Model(inputs=inputs, outputs=outputs)
+    return model
+
+
+wide_and_deep_model = create_wide_and_deep_model()
+
+keras.utils.plot_model(wide_and_deep_model, show_shapes=True)
+
+run_experiment(wide_and_deep_model)
+
+"""
+The wide and deep model achieves ~79.8% test accuracy.
+"""
+
+"""
+## Experiment 3: deep and cross model
+
+In the third experiment, we create a deep and cross model. The deep part of this model is the same as
+the deep part created in the previous experiment. The key idea of the cross part is to apply explicit feature
+crossing in an efficient way, where the degree of cross features grows with layer depth.
```

Comment:
Have you considered using a `CategoryCrossing` layer? It exists precisely for this purpose. https://keras.io/api/layers/preprocessing_layers/categorical/category_crossing/

Reply:
I tried to use it, but it seems that it doesn't handle the produced "crossed" vocab automatically. That is, in order to generate indices for the output of the category crossing layer, I need to *manually* generate all the cross combinations, to create a `StringLookup` . More precisely, this is what I think I need to do to use `CategoryCrossing`:

```
crossing_vocab = my_create_crossing_vocab_fn(dataframe_column_a, datafram_column_b)
index = StringLookup(vocabulary=crossing_vocab)
category_crossing_layer = CategoryCrossing()
crossed_categories = category_crossing_layer([a, b])
crossed_category_indicies = index(crossed_categories)
....
```
But I think I can simply do:

```
dataframe.a_x_b = my_create_crossing_column_fn(dataframe_column_a, datafram_column_b)
...
index = StringLookup(vocabulary=list(dataframe.a_x_b.unique))
crossed_category_indicies = index(inputs['a_x_b'])
...
```

Thus, I am not sure about the benefit of it. Please LMK if I am missing something...

Reply:
Sounds good, we can leave the code as-is.